proposals/NNNN-importing-forward-declared-objc-interfaces-and-protocols.md

```diff
@@ -0,0 +1,177 @@
+class Foo : NSObject {}
+
+// @protocol Bar turns into
+@available(*, unavailable, message: “This Objective-C protocol has only been forward declared; import its owning module to use it”)
+protocol Bar : NSObjectProtocol {}
+```
+
+The idea is to introduce the minimal change that will make Objective-C APIs usable in a predictable safe manner.
+
+The aforementioned Objective-C API with this change looks like this from Swift:
+
+```swift
+@available(*, unavailable, message: "This Objective-C class has only been forward-declared; import its owning module to use it")
+class ForwardDeclaredInterface {
+}
+@available(*, unavailable, message: "This Objective-C protocol has only been forward-declared; import its owning module to use it")
+protocol ForwardDeclaredProtocol : NSObjectProtocol {
+}
+class IncompleteTypeConsumer1 : NSObject {
+  var propertyUsingAForwardDeclaredProtocol1: ForwardDeclaredProtocol!
+  var propertyUsingAForwardDeclaredInterface1: ForwardDeclaredInterface!
+  init!()
+  func methodReturningForwardDeclaredProtocol1() -> ForwardDeclaredProtocol!
+  func methodReturningForwardDeclaredInterface1() -> ForwardDeclaredInterface!
+  func methodTakingAForwardDeclaredProtocol1(_ param: ForwardDeclaredProtocol!)
+  func methodTakingAForwardDeclaredInterface1(_ param: ForwardDeclaredInterface!)
+}
+func CFunctionReturningAForwardDeclaredInterface1() -> ForwardDeclaredInterface!
+func CFunctionTakingAForwardDeclaredInterface1(_ param: ForwardDeclaredInterface!)
+func CFunctionReturningAForwardDeclaredProtocol1() -> ForwardDeclaredProtocol!
+func CFunctionTakingAForwardDeclaredProtocol1(_ param: ForwardDeclaredProtocol!)
+```
+
+More usage examples can be found in these tests introduced here: https://github.com/apple/swift/pull/61606
+
+## Detailed design
+
+Modifications lie almost exclusively in the ClangImporter -- specifically in the SwiftDeclConverter. If asked to convert a
+`clang::ObjCInterfaceDecl` or `clang::ObjCProtocolDecl` with no definition the SwiftDeclConverter will now return a placeholder type instead of
+bailing. These placeholder types are as described:
+
+```swift
+// @class Foo turns into
+@available(*, unavailable, message: “This Objective-C class has only been forward declared; import its owning module to use it”)
+class Foo : NSObject {}
+
+// @protocol Bar turns into
+@available(*, unavailable, message: “This Objective-C protocol has only been forward declared; import its owning module to use it”)
+protocol Bar : NSObjectProtocol {}
+```
+
+Permitted usages of these types are intentionally limited. You will be able to use Objective-C and C declarations that refer to these types without issue.
+You will be able to pass around instances of these incomplete types from Swift to Objective-C and vice versa.
+You’ll also be able to call the set of methods provided by NSObject or NSObjectProtocol on instances.
```

Comment:
```suggestion
You’ll also be able to call the set of methods provided by `NSObject` or `NSObjectProtocol` on instances.
```

My Objective-C is a bit rusty these days; is this also possible on pointers to only-forward-declared class/protocol types today in that language? Does it assume `NSObject(Protocol)` inheritance rather than the possibility of some other root class?

Reply:
It does make this assumption, but as I understand it Objective-C interop in general already makes this assumption. I'm told that Swift does not know how to interact with objects that don't inherit from NSObject.

I tried a simple example with the release compiler and it crashed at runtime:
```
./src/swift-client-complete
2023-01-04 12:34:03.220 swift-client-complete[43116:24260633] *** NSForwarding: warning: object 0x1041740c0 of class 'Foo' does not implement methodSignatureForSelector: -- trouble ahead
2023-01-04 12:34:03.221 swift-client-complete[43116:24260633] *** NSForwarding: warning: object 0x1041740c0 of class 'Foo' does not implement doesNotRecognizeSelector: -- abort
Trace/BPT trap: 5
```

In general I believe if we are wrong about the assumptions, you'll get a message send failure crash at runtime. I've been assured that NSObject is required for interop, but I'm not sure about `NSObjectProtocol`. It would be helpful to get some feedback from those familiar with the details here. In any case, being able to call the methods provided by `NSObject` and `NSObjectProtocol` is more of a nicety, we can drop that if it turns out to be unsound.